src/core/thread/mle.cpp

```diff
@@ -1446,24 +1448,40 @@ otError Mle::AppendAddressRegistration(Message &aMessage, AddressRegistrationMod
         VerifyOrExit(counter < OPENTHREAD_CONFIG_MLE_IP_ADDRS_TO_REGISTER, OT_NOOP);
     }
 
-    // For sleepy end device, register external multicast addresses to the parent for indirect transmission
-    if (!IsRxOnWhenIdle())
+    // Append external multicast address.
+    // For sleepy end device, register external multicast addresses to the parent for indirect transmission.
+    // Since Thread 1.2, non-sleepy MED would depend on its Thread 1.2 parent as the proxy to perform
+    // Multicast Listener Report, it registers external multicast addresses of scope larger than realm
+    // local as well.
+    while (Get<ThreadNetif>().GetNextExternalMulticast(iterator, address) == OT_ERROR_NONE)
```

Comment:
We can use `ShouldRegisterMulticastToParent` to further simplify this code make it it more readable. 
```suggestion

    VerifyOrExit(ShouldRegisterMulticastToParent(), OT_NOOP);

    while (Get<ThreadNetif>().GetNextExternalMulticast(iterator, address) == OT_ERROR_NONE)
    {
#if (OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2)
        if (IsRxOnWhenIdle() && !address.IsMulticastLargerThanRealmLocal())
        {
            // For Thread 1.2 MED, skip multicast address with scope not larger than realm local when registering.
            continue;
        }
#endif

        entry.SetUncompressed();
        entry.SetIp6Address(address);
        SuccessOrExit(error = aMessage.Append(&entry, entry.GetLength()));
        length += entry.GetLength();

        counter++;
        // only continue to append if there is available entry.
        VerifyOrExit(counter < OPENTHREAD_CONFIG_MLE_IP_ADDRS_TO_REGISTER, OT_NOOP);
    }

#endif

```

Reply:
Note that `Mle::AppendAddressRegistration()` itself is an indication that it is MTD end device, no need to perform the below check in `Mle::ShouldRegisterMulticastToParent()`

https://github.com/openthread/openthread/blob/210223e9af2e748f41c72ec5cf7de6bf2b0fb1e0/src/core/thread/mle.cpp#L4493-L4497

When it goes to here, the only decision to make is whether or not to append the multicast address. As for readability, I think per situation is well commented in existing code, which seems to me more clear to understand when reading.

https://github.com/openthread/openthread/blob/210223e9af2e748f41c72ec5cf7de6bf2b0fb1e0/src/core/thread/mle.cpp#L1458-L1474

Prefer to keep it as it.

Reply:
> Note that `Mle::AppendAddressRegistration()` itself is an indication that it is MTD end device, no need to perform the below check in `Mle::ShouldRegisterMulticastToParent()`

Yes, `Mle::ShouldRegisterMulticastToParent()` do some redundant checks here. 

> For `Mle::AppendAddressRegistration()`, when it goes here, the only decision to make is whether or not to append the multicast address. As for readability, I think per situation is well commented in existing code, which seems to me more clear to understand when reading.
> 
> https://github.com/openthread/openthread/blob/210223e9af2e748f41c72ec5cf7de6bf2b0fb1e0/src/core/thread/mle.cpp#L1458-L1474
> 

The code here does same checks of `Mle::ShouldRegisterMulticastToParent()`:
- Skip 1.1 MED
- Skip 1.2 MED with 1.1 parent

So, I would suggest to use `Mle::ShouldRegisterMulticastToParent()` for better code reuse. 

Besides, after `VerifyOrExit(ShouldRegisterMulticastToParent(), OT_NOOP);`, we are facing only 2 conditions:
1. SED
2. 1.2 MED with 1.2 Parent
This can help code readers easier to understand the code?

Reply:
Personally prefer existing code a bit, per condition here well matches the proceeding summary comment when appending external multicast address - which seems to me more friendly to understand.

https://github.com/openthread/openthread/blob/210223e9af2e748f41c72ec5cf7de6bf2b0fb1e0/src/core/thread/mle.cpp#L1451-L1455

Reply:
Fine with current code since it's only personal preference.